components/TextFields/src/MDCTextInputController.m

```diff
@@ -0,0 +1,1199 @@
+
+#pragma mark - Underline Labels Fonts
+
++ (UIFont *)placeholderFont {
+  return [UIFont mdc_preferredFontForMaterialTextStyle:MDCFontTextStyleBody1];
+}
+
++ (UIFont *)underlineLabelsFont {
+  return [UIFont mdc_preferredFontForMaterialTextStyle:MDCFontTextStyleCaption];
+}
+
+#pragma mark - Properties Implementation
+
+- (void)setCharacterCountViewMode:(UITextFieldViewMode)characterCountViewMode {
+  if (_characterCountViewMode != characterCountViewMode) {
+    _characterCountViewMode = characterCountViewMode;
+
+    [self updateLayout];
+  }
+}
+
+- (void)setErrorAccessibilityValue:(NSString *)errorAccessibilityValue {
+  _errorAccessibilityValue = [errorAccessibilityValue copy];
+}
+
+- (void)setErrorColor:(UIColor *)errorColor {
+  if (![_errorColor isEqual:errorColor]) {
+    _errorColor = errorColor;
+    if (self.isDisplayingErrorText) {
+      [self updateLeadingUnderlineLabel];
+      [self updateTrailingUnderlineLabel];
+      [self updateUnderline];
+    }
+  }
+}
+
+- (void)setErrorText:(NSString *)errorText {
+  _errorText = [errorText copy];
+}
+
+- (void)setFloatingPlaceholderColor:(UIColor *)floatingPlaceholderColor {
+  if (![_floatingPlaceholderColor isEqual:floatingPlaceholderColor]) {
+    _floatingPlaceholderColor = floatingPlaceholderColor;
+    [self updatePlaceholder];
+  }
+}
+
+- (void)setFloatingPlaceholderScale:(NSNumber *)floatingPlaceholderScale {
+  if (![_floatingPlaceholderScale isEqualToNumber:floatingPlaceholderScale]) {
+    _floatingPlaceholderScale = floatingPlaceholderScale;
+    [self updatePlaceholder];
+  }
+}
+
+- (void)setHelperText:(NSString *)helperText {
+  if (self.isDisplayingErrorText) {
+    self.previousLeadingText = [helperText copy];
+  } else {
+    if (![self.textInput.leadingUnderlineLabel.text isEqualToString:helperText]) {
+      self.textInput.leadingUnderlineLabel.text = helperText;
+      [self updateLayout];
+    }
+  }
+}
+
+- (NSString *)helperText {
+  if (self.isDisplayingErrorText) {
+    return self.previousLeadingText;
+  } else {
+    return self.textInput.leadingUnderlineLabel.text;
+  }
+}
+
+- (void)setInlinePlaceholderColor:(UIColor *)inlinePlaceholderColor {
+  if (![_inlinePlaceholderColor isEqual:inlinePlaceholderColor]) {
+    _inlinePlaceholderColor = inlinePlaceholderColor;
+    [self updatePlaceholder];
+  }
+}
+
+- (BOOL)isDisplayingCharacterCountError {
+  return self.characterCountMax && [self characterCount] > self.characterCountMax;
+}
+
+- (BOOL)isDisplayingErrorText {
+  return self.errorText != nil;
+}
+
+- (void)setPresentationStyle:(MDCTextInputPresentationStyle)presentationStyle {
+  if (_presentationStyle != presentationStyle) {
+    _presentationStyle = presentationStyle;
+
+    BOOL isDirectionToUp = NO;
+    if (presentationStyle == MDCTextInputPresentationStyleFloatingPlaceholder) {
+      isDirectionToUp = self.textInput.text.length > 1 || self.textInput.isEditing;
+    }
+
+    [CATransaction begin];
+    [CATransaction setAnimationDuration:0];
```

Comment:
I dont' want to remove the transaction because I want to make sure all of that code happens in the same pass of the render server.